### PR TITLE
Reduce helix controller log and minor code improve

### PR DESCRIPTION
**Pull request by qqu0127**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:

Fix https://github.com/apache/helix/issues/2080

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:
We are seeing issue in certain product where controller is generating way too many logs that fill up disk space. 
In this PR, we reduce a few logger level from INFO to DEBUG and update the logic to avoid duplicated messages in `WorkflowDispatcher` (Step 5: handle workflow that should STOP)

### Tests

- [X] The following tests are written for this issue:

[info] ./zookeeper-api/target/surefire-reports/TestSuite.txt: Tests run: 54, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 123.232 s - in TestSuite
[info] ./recipes/rsync-replicated-file-system/target/surefire-reports/TestSuite.txt: Tests run: 0, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.585 s - in TestSuite
[info] ./recipes/task-execution/target/surefire-reports/TestSuite.txt: Tests run: 0, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.582 s - in TestSuite
[info] ./recipes/distributed-lock-manager/target/surefire-reports/TestSuite.txt: Tests run: 0, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.705 s - in TestSuite
[info] ./recipes/rabbitmq-consumer-group/target/surefire-reports/TestSuite.txt: Tests run: 0, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.78[9](https://github.com/apache/helix/runs/6494616692?check_suite_focus=true#step:6:10) s - in TestSuite
[info] ./recipes/service-discovery/target/surefire-reports/TestSuite.txt: Tests run: 0, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.415 s - in TestSuite
[info] ./metadata-store-directory-common/target/surefire-reports/TestSuite.txt: Tests run: 31, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.745 s - in TestSuite
[info] ./helix-rest/target/surefire-reports/TestSuite.txt: Tests run: 205, Failures: 1, Errors: 0, Skipped: 35, Time elapsed: 138.881 s <<< FAILURE! - in TestSuite
Error:  Test failed: testGetClusters(org.apache.helix.rest.server.TestClusterAccessor)  Time elapsed: 0.513 s  <<< FAILURE!
[info] ./metrics-common/target/surefire-reports/TestSuite.txt: Tests run: 0, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.335 s - in TestSuite
[info] ./helix-view-aggregator/target/surefire-reports/TestSuite.txt: Tests run: 13, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 58.735 s - in TestSuite
[info] ./helix-common/target/surefire-reports/TestSuite.txt: Tests run: 0, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.272 s - in TestSuite
[info] ./helix-core/target/surefire-reports/TestSuite.txt: Tests run: 1314, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 5,903.158 s - in TestSuite
[info] ./helix-lock/target/surefire-reports/TestSuite.txt: Tests run: [11](https://github.com/apache/helix/runs/6494616692?check_suite_focus=true#step:6:12), Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 53.896 s - in TestSuite

Known issue with `TestClusterAccessor`, verified passed locally.

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/task/assigner/ThreadCountBasedTaskAssigner.java [hunk 1]
```diff
@@ -27,6 +27,7 @@
 import java.util.Map;
 import java.util.PriorityQueue;
 
+import java.util.Set;
```
Comment (desaikomal) [inline helix-core/src/main/java/org/apache/helix/task/assigner/ThreadCountBasedTaskAssigner.java:30]: is this intended part of the change?
As this doesn't look like log level related change

File: helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java [hunk 2]
```diff
@@ -150,7 +149,10 @@ && isTimeout(workflowCtx.getStartTime(), workflowCfg.getTimeout())) {
...
     if (!TaskConstants.FINAL_STATES.contains(workflowCtx.getWorkflowState())
         && TargetState.STOP.equals(targetState)) {
-      LOG.info("Workflow {} is marked as stopped. Workflow state is {}", workflow,
+      if (workflowCtx.getWorkflowState() == TaskState.STOPPED) {
+        return;
+      }
+      LOG.debug("Workflow {} is marked as stopped. Workflow state is {}", workflow,
           workflowCtx.getWorkflowState());
       if (isWorkflowStopped(workflowCtx, workflowCfg)) {
         workflowCtx.setWorkflowState(TaskState.STOPPED);
```
Comment (qqu0127) [inline helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java:158]: Justification: 
Ideally the message `Workflow {} is marked as stopped` should only be logged once, same for `workflowCtx.setWorkflowState`, but it's triggered for each task in the workflow. Verified there are tons of messages saying `Workflow ABC is marked as stopped. Workflow state is STOPPED`, meaning the workflow has been stopped already.
The only other place that set state STOPPED is in `JobDispatcher` where the rest of the update logic is handled. As a result, early termination (if state == TaskState.STOPPED) makes sense here.

File: helix-core/src/main/java/org/apache/helix/task/assigner/ThreadCountBasedTaskAssigner.java [hunk 3]
```diff
@@ -74,19 +75,16 @@ public Map<String, TaskAssignResult> assignTasks(Iterable<AssignableInstance> as
...
+      assignableInstances.add(assignableInstanceManager.getAssignableInstance(instance));
     }
 
     if (tasks == null || !tasks.iterator().hasNext()) {
-      logger.warn("No task to assign!");
       return Collections.emptyMap();
     }
-    if (assignableInstances == null || !assignableInstances.iterator().hasNext()) {
-      logger.warn("No instance to assign!");
+    if (assignableInstances.isEmpty()) {
```
Comment (junkaixue) [inline helix-core/src/main/java/org/apache/helix/task/assigner/ThreadCountBasedTaskAssigner.java:87]: What if it is null?

File: helix-core/src/main/java/org/apache/helix/task/assigner/ThreadCountBasedTaskAssigner.java [hunk 4]
```diff
@@ -74,19 +75,16 @@ public Map<String, TaskAssignResult> assignTasks(Iterable<AssignableInstance> as
...
-          .add(assignableInstanceManager.getAssignableInstance(instance));
+      assignableInstances.add(assignableInstanceManager.getAssignableInstance(instance));
     }
 
     if (tasks == null || !tasks.iterator().hasNext()) {
-      logger.warn("No task to assign!");
       return Collections.emptyMap();
     }
-    if (assignableInstances == null || !assignableInstances.iterator().hasNext()) {
-      logger.warn("No instance to assign!");
```
Comment (junkaixue) [inline helix-core/src/main/java/org/apache/helix/task/assigner/ThreadCountBasedTaskAssigner.java:89]: If we remove this, how to determine all quota full or something?

Comment (junkaixue) [inline helix-core/src/main/java/org/apache/helix/task/assigner/ThreadCountBasedTaskAssigner.java:89]: If we remove this, how to determine all quota full or something?

File: helix-core/src/main/java/org/apache/helix/task/assigner/ThreadCountBasedTaskAssigner.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (qqu0127) [reply helix-core/src/main/java/org/apache/helix/task/assigner/ThreadCountBasedTaskAssigner.java:87]: It's never null, see `Set<AssignableInstance> assignableInstances = new HashSet<>();` in line 78

File: helix-core/src/main/java/org/apache/helix/controller/stages/MessageGenerationPhase.java [hunk 5]
```diff
@@ -320,18 +320,18 @@ private Message generateCancellationMessageForPendingMessage(final String desire
     if (pendingMessage != null) {
       String pendingState = pendingMessage.getToState();
       if (nextState.equalsIgnoreCase(pendingState)) {
-        LogUtil.logInfo(logger, _eventId,
+        LogUtil.logDebug(logger, _eventId,
```
[the remark this answers is not in the document]
Comment (qqu0127) [reply helix-core/src/main/java/org/apache/helix/controller/stages/MessageGenerationPhase.java:323]: I see, thanks. Is there other log that you think we should better keep INFO?

File: helix-core/src/main/java/org/apache/helix/task/assigner/ThreadCountBasedTaskAssigner.java (hunk 4 above)
Comment (qqu0127) [reply helix-core/src/main/java/org/apache/helix/task/assigner/ThreadCountBasedTaskAssigner.java:89]: Yeah, that's the dilemma. I'm trying to move the check and log to the build up phase in `AssignableInstanceManager`. I think we don't have to print this for every task assignment.

**Review by desaikomal — COMMENTED** [review]
LGTM, Quincy.

File: helix-core/src/main/java/org/apache/helix/task/AssignableInstanceManager.java [hunk 6]
```diff
@@ -178,6 +178,9 @@ public void buildAssignableInstances(ClusterConfig clusterConfig, TaskDataCache
     }
     LOG.info(
         "AssignableInstanceManager built AssignableInstances from scratch based on contexts in TaskDataCache due to Controller switch or ClusterConfig change.");
+    if (_assignableInstanceMap.isEmpty()) {
+      LOG.warn("No assignable instance!");
```
Comment (xyuanlu) [inline helix-core/src/main/java/org/apache/helix/task/AssignableInstanceManager.java:182]: May I suggest let's add more info here? So we could differentiate between the 3 logs here.

Comment (qqu0127) [reply helix-core/src/main/java/org/apache/helix/task/AssignableInstanceManager.java:182]: Thanks for the review.
The above INFO log should provide enough context to differentiate 3 cases. I think it's fine to keep it this way. Let me know if you are fine with it. Thanks

**Comment by qqu0127** [comment]
This PR is ready to merge, approved by @desaikomal 
Commit message:
Reduce helix controller log and minor code improve

File: helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java (hunk 2 above)
Comment (NealSun96) [reply helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java:158]: @qqu0127 I suggest moving this condition into the above IF.

File: helix-core/src/main/java/org/apache/helix/task/assigner/ThreadCountBasedTaskAssigner.java (hunk 4 above)
Comment (NealSun96) [reply helix-core/src/main/java/org/apache/helix/task/assigner/ThreadCountBasedTaskAssigner.java:89]: `handleAdditionalTaskAssignment` is called once per pipeline, isn't it? Is this log a problem?

File: helix-core/src/main/java/org/apache/helix/task/assigner/ThreadCountBasedTaskAssigner.java [hunk 7]
```diff
@@ -74,19 +75,16 @@ public Map<String, TaskAssignResult> assignTasks(Iterable<AssignableInstance> as
   public Map<String, TaskAssignResult> assignTasks(
       AssignableInstanceManager assignableInstanceManager, Collection<String> instances,
       Iterable<TaskConfig> tasks, String quotaType) {
-    Iterable<AssignableInstance> assignableInstances = new HashSet<>();
+    Set<AssignableInstance> assignableInstances = new HashSet<>();
     // Only add the AssignableInstances that are also in instances
     for (String instance : instances) {
-      ((HashSet<AssignableInstance>) assignableInstances)
-          .add(assignableInstanceManager.getAssignableInstance(instance));
+      assignableInstances.add(assignableInstanceManager.getAssignableInstance(instance));
```
Comment (NealSun96) [inline helix-core/src/main/java/org/apache/helix/task/assigner/ThreadCountBasedTaskAssigner.java:81]: Looks like the below check of `assignableInstances` can be moved here - not that it helps with logs, but the logic can be simpliefied.

**Review by NealSun96 — CHANGES_REQUESTED** [review]

File: helix-core/src/main/java/org/apache/helix/task/assigner/ThreadCountBasedTaskAssigner.java (hunk 7 above)
Comment (qqu0127) [reply helix-core/src/main/java/org/apache/helix/task/assigner/ThreadCountBasedTaskAssigner.java:81]: You mean check on `assignableInstances.isEmpty()`? The logic will be wrong if move into the forloop. Am I missing anything?

Comment (NealSun96) [reply helix-core/src/main/java/org/apache/helix/task/assigner/ThreadCountBasedTaskAssigner.java:81]: Synced offline: this is a NIT.

File: helix-core/src/main/java/org/apache/helix/task/AssignableInstanceManager.java [hunk 8]
```diff
@@ -419,7 +425,9 @@ public void updateAssignableInstances(ClusterConfig clusterConfig,
     }
     LOG.info(
         "AssignableInstanceManager updated AssignableInstances due to LiveInstance/InstanceConfig change.");
-
+    if (_assignableInstanceMap.isEmpty()) {
+      LOG.warn("No assignable instance!");
```
Comment (NealSun96) [inline helix-core/src/main/java/org/apache/helix/task/AssignableInstanceManager.java:429]: Synced offline: these can be removed.

**Review by NealSun96 — CHANGES_REQUESTED** [review]

**Comment by qqu0127** [comment]
Verified the test passed locally, this is ready to merge @NealSun96 thanks for the review
Commit message:
Reduce helix controller log and minor code improve
Turn down a few log level to DEBUG and reduce duplicated logs

[INFO] Tests run: 39, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 229.346 s - in org.apache.helix.rest.server.TestClusterAccessor
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 39, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- jacoco-maven-plugin:0.8.6:report (generate-code-coverage-report) @ helix-rest ---
[INFO] Loading execution data file /Users/qqu/workspace/qqu-helix/helix-rest/target/jacoco.exec
[INFO] Analyzed bundle 'Apache Helix :: Restful Interface' with 91 classes
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  03:56 min
[INFO] Finished at: 2022-05-26T15:15:37-07:00
[INFO] ------------------------------------------------------------------------

**Review by NealSun96 — APPROVED** [review]

**Pull request merged**